lib/matplotlib/colorbar.py

```diff
@@ -675,7 +720,30 @@ def add_lines(self, levels, colors, linewidths, erase=True):
             for each line.
         erase : bool, default: True
             Whether to remove any previously added lines.
+
+        Notes
+        -----
+        For back-compatibility, this method can also be called with the
+        signature ``colorbar.add_lines(contour_set, erase=True)``, in which
+        case *levels*, *colors*, and *linewidths* are taken from *contour_set*.
         """
+        params = _api.select_matching_signature(
+            [lambda self, CS, erase=True: locals(),
+             lambda self, levels, colors, linewidths, erase=True: locals()],
+            self, *args, **kwargs)
+        if "CS" in params:
+            self, CS, erase = params.values()
+            # Old compat API for Colorbar.add_lines taking a ContourSet.
+            if not isinstance(CS, contour.ContourSet) or CS.filled:
+                raise ValueError("If a single artist is passed to add_lines, "
+                                 "it must be a ContourSet of lines")
+            # TODO: Make colorbar lines auto-follow changes in contour lines.
+            levels = CS.levels
+            colors = [c[0] for c in CS.tcolors]
+            linewidths = [t[0] for t in CS.tlinewidths]
+        else:
+            self, levels, colors, linewidths, erase = params.values()
```

Comment:
I'm a bit confused that this path isn't covered by the tests... Above you say the other one is the deprecated signature, but it appears to be the only one used, so should this path be the deprecated signature instead?

Reply:
Yeah, good point - if we are gong to have the funky signature matching they should probably be tested....

Reply:
Good points; I went back to the old approach (of having one of the signatures forward to the other), so coverage should be fine now; and I tweaked the wording to not have to decide which of the signatures (if any) we really prefer.  (I also removed the `__signature__` patching.)  We can always decide on the preferred signature in a followup PR...